src/uu/install/src/install.rs

```diff
@@ -748,29 +751,70 @@ fn perform_backup(to: &Path, b: &Behavior) -> UResult<Option<PathBuf>> {
 /// Returns an empty Result or an error in case of failure.
 ///
 fn copy_file(from: &Path, to: &Path) -> UResult<()> {
+    // open the source file and get a file descriptor, making sure the source file exists
+    #[cfg(unix)]
+    let src = fs::OpenOptions::new().read(true).open(from);
+    #[cfg(not(unix))]
+    let src = File::open(from);
+    if let Err(e) = src {
+        show_error!(
+            "Failed to open source file {}. Error: {:?}",
+            from.display(),
+            e
+        );
+        return Err(InstallError::InstallFailed(from.to_path_buf(), to.to_path_buf(), e).into());
+    }
+    let mut src = src.unwrap();
+
     // fs::copy fails if destination is a invalid symlink.
     // so lets just remove all existing files at destination before copy.
-    if let Err(e) = fs::remove_file(to) {
-        if e.kind() != std::io::ErrorKind::NotFound {
-            show_error!(
-                "Failed to remove existing file {}. Error: {:?}",
-                to.display(),
-                e
-            );
+    let remove_destination = || {
+        if let Err(e) = fs::remove_file(to) {
+            if e.kind() != std::io::ErrorKind::NotFound {
+                show_error!(
+                    "Failed to remove existing file {}. Error: {:?}",
+                    to.display(),
+                    e
+                );
+            }
         }
+    };
+    remove_destination();
+
+    // create the destination file first. Using safer mode on unix to avoid
+    // potential unsafe mode between time-of-creation and time-of-chmod.
+    #[cfg(unix)]
+    let dst = fs::OpenOptions::new()
+        .write(true)
+        .create_new(true)
+        .mode(0o600)
+        .open(to);
+    #[cfg(not(unix))]
+    let dst = File::create(to);
+
+    if let Err(e) = dst {
+        show_error!(
+            "Failed to create destination file {}. Error: {:?}",
+            to.display(),
+            e
+        );
+        return Err(InstallError::InstallFailed(from.to_path_buf(), to.to_path_buf(), e).into());
     }
-
-    if from.as_os_str() == "/dev/null" {
-        /* workaround a limitation of fs::copy
-         * https://github.com/rust-lang/rust/issues/79390
-         */
-        if let Err(err) = File::create(to) {
+    let mut dst = dst.unwrap();
+
+    /* workaround a limitation of fs::copy: skip copy if source is /dev/null
+     * https://github.com/rust-lang/rust/issues/79390
+     */
+    if from.as_os_str() != "/dev/null" {
+        // use `std::io::copy` to avoid unsafe file modes
+        if let Err(err) = std::io::copy(&mut src, &mut dst) {
+            drop(src);
+            drop(dst);
```

Comment:
are these drops necessary? unlink should proceed just fine at least on unix systems, no idea about windows

Reply:
These two early drops are for closing two file descriptors before `unlink`. Technically the `unlink` will succeed even when there are opened fds... WDYT?